zebra-consensus/src/block/subsidy/founders_reward.rs

```diff
@@ -7,17 +7,25 @@ use std::convert::TryFrom;
 use zebra_chain::{
     amount::{Amount, Error, NonNegative},
     block::Height,
-    parameters::Network,
+    parameters::{Network, NetworkUpgrade::*},
+    transaction::Transaction,
+    transparent::{Address, Output},
 };
 
 use crate::block::subsidy::general::{block_subsidy, halving_divisor};
-use crate::parameters::subsidy::FOUNDERS_FRACTION_DIVISOR;
+use crate::parameters::subsidy::{
+    BLOSSOM_POW_TARGET_SPACING_RATIO, FOUNDERS_FRACTION_DIVISOR, FOUNDERS_REWARD_ADDRESSES_MAINNET,
+    FOUNDERS_REWARD_ADDRESSES_TESTNET, FOUNDER_ADDRESS_CHANGE_INTERVAL,
+};
 
 /// `FoundersReward(height)` as described in [protocol specification §7.7][7.7]
 ///
 /// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
 pub fn founders_reward(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
-    if halving_divisor(height, network) == 1 {
+    let canopy_height = Canopy
+        .activation_height(network)
+        .expect("canopy activation height should be available");
+    if halving_divisor(height, network) == 1 && height < canopy_height {
```

Comment:
Checking only the halving is not enough for the testnet so looking for halving and canopy now to return non zero amounts. This was done while making the tests work for the testnet at https://github.com/ZcashFoundation/zebra/pull/1170/commits/01f1007b6c955e75d04071ef44ebf08a193b7f8f

Reply:
I fixed this bug in `subsidy_is_valid` in #1051, but forgot to fix the other places it was used.

We're now doing this check in at least 3 places, in 3 slightly different ways:
* `subsidy_is_valid`
* `founders_reward` (amount function)
* `founders_reward_address`

Instead, let's define a `founders_reward_active` function, and use that in all 3 places.

I suggest we use the code from `subsidy_is_valid`, because it is the simplest implementation of the consensus rule: "founders rewards are active until canopy activation".

https://github.com/ZcashFoundation/zebra/blob/main/zebra-consensus/src/block/check.rs#L112

Reply:
Good idea, i made this at https://github.com/ZcashFoundation/zebra/pull/1170/commits/57e188fc48bba3fc410f5948eb8f199c09e8eb16

Reply:
Founders' Rewards happen to be paid until Canopy activation, but that isn't the rule. The rule is that they are paid until Canopy activation of the first halving, whichever is first. You could assert that the Canopy activation height is set and is no later than the first having, and *then* it would be sufficient to check "before Canopy activation".

Reply:
The checks are all done in a single function now, so we just have to change them in one place. That was the point of my original review and fix PR.

The consensus rules are explained differently in different specifications. That's probably how we ended up with different checks.

Are these specifications equivalent on mainnet and testnet?
If they aren't, what sort of edge cases should we be looking out for?

Zcash Specification, section 3.9:
[Pre-Canopy] The block subsidy is composed of a miner subsidy and a Founders’ Reward.
[Canopy onward] The block subsidy is composed of a miner subsidy and a series of funding streams.

ZIP 207:
Once the Canopy network upgrade activates:
The existing consensus rule for payment of the Founders' Reward 8 is no longer active. (This would be the case under the preexisting consensus rules for Mainnet, but not for Testnet.)

ZIP 214:
Since ZIP 1014 specifies that the Zcash Development Fund starts at the first halving, the activation height of Canopy on Mainnet therefore SHALL be 1046400.

ZIP 214 also specifies the exact start and end heights  for funding streams: https://zips.z.cash/zip-0214#specification

ZIP 1014:
Starting at the first Zcash halving in 2020, until the second halving in 2024, 20% of the block subsidy of each block SHALL be allocated to a "Dev Fund" that consists of the following three slices

It appears that ZIP 1014 only applies to mainnet, but that's not explicitly stated in the document.

Reply:
So i added the halving check back. Now into the `founders_reward_active()` function in https://github.com/ZcashFoundation/zebra/pull/1170/commits/0008a3fc5c6a7ff7b4330178b445ae090be51f8c.

Reply:
This spec ambiguity is now tracked in #1221.